3rdParty/fuerte/src/H1Connection.cpp

```diff
@@ -222,33 +224,54 @@ size_t H1Connection<ST>::requestsLeft() const {
   return q;
 }
 
+template <SocketType ST>
+void H1Connection<ST>::activate() {
+  FUERTE_ASSERT(_active.load());
+  Connection::State state = this->_state.load();
+  if (state == Connection::State::Connected) {
+    FUERTE_LOG_HTTPTRACE << "activate: connected\n";
+    startWriting();
+  } else if (state == Connection::State::Disconnected) {
+    FUERTE_LOG_HTTPTRACE << "activate: not connected\n";
+    this->startConnection();
+  } else if (state == Connection::State::Failed) {
+    FUERTE_LOG_ERROR << "activate: queued request on failed connection\n";
+    drainQueue(fuerte::Error::ConnectionClosed);
+    _active.store(false);    // No more activity from our side
+  }
+  // If the state is `Connecting`, we do not need to do anything, this can
+  // happen if this `activate` was posted when the connection was still
+  // `Disconnected`, but in the meantime the connect has started.
+}
+
 template <SocketType ST>
 void H1Connection<ST>::finishConnect() {
+  // Note that the connection timeout alarm has already been disarmed.
+  // If it has already gone off, we
   auto exp = Connection::State::Connecting;
   if (this->_state.compare_exchange_strong(exp, Connection::State::Connected)) {
     startWriting();  // starts writing queue if non-empty
+  } else {
+    FUERTE_LOG_ERROR << "finishConnect: found state other than 'Connecting'";
+    FUERTE_ASSERT(false);
+    // If this happens, we probably have a sleeping barber
   }
 }
 
-// Thread-Safe: activate the combined write-read loop
+// startWriting must only be executed on the IO thread
 template <SocketType ST>
 void H1Connection<ST>::startWriting() {
-  FUERTE_LOG_HTTPTRACE << "startWriting: this=" << this << "\n";
-  if (!_active) {
-    this->_io_context->post([self(Connection::shared_from_this())] {
-      auto& me = static_cast<H1Connection<ST>&>(*self);
-      FUERTE_LOG_HTTPTRACE << "startWriting: active=true, this=" << &me << "\n";
-      if (!me._active.exchange(true)) {  // we are the only ones here now
-        // we might get in a race with shutdownConnection()
-        Connection::State state = me._state.load();
-        if (state != Connection::State::Connected) {
-          me._active.store(false);
-          me.startConnection();
-        } else {
-          me.asyncWriteNextRequest();
-        }
-      }
-    });
+  FUERTE_ASSERT(_active.load());
```

Comment:
why is this code the same in activate()

Reply:
You are right, this is not yet good. Will fix. I am currently converging model and code and this is one more thing to fix.

Reply:
Fixed.